File: bindings/csharp/Sop.Examples/BtreeBasic.cs
using System;
using System.Collections.Generic;
using Sop;

namespace Sop.Examples;

public static class BtreeBasic
{
    public static void Run()
    {
        Console.WriteLine("\n--- Running Basic B-Tree Operations ---");

        using var ctx = new Context();
        
        // 1. Open Database
        var dbOpts = new DatabaseOptions 
        { 
            StoresFolders = new List<string> { "sop_data_basic" },
            Type = (int)DatabaseType.Standalone
        };
        var db = new Database(dbOpts);

        // 2. Start Transaction
        using var trans = db.BeginTransaction(ctx);
        try
        {
            // 3. Create/Open B-Tree
            var btree = db.NewBtree<string, string>(ctx, "users_basic", trans);

            // 4. Add Items (Create)
            Console.WriteLine("Adding users...");
            btree.Add(ctx, new Item<string, string>("user1", "Alice"));
            btree.Add(ctx, new Item<string, string>("user2", "Bob"));
            btree.Add(ctx, new Item<string, string>("user3", "Charlie"));

            // 5. Find & Get (Read)
            if (btree.Find(ctx, "user1"))
            {
                var items = btree.GetValues(ctx, "user1");
                Console.WriteLine($"Found: {items[0].Key} -> {items[0].Value}");
            }

            // 6. Update
            Console.WriteLine("Updating user2...");
            btree.Update(ctx, new List<Item<string, string>> 
            { 
                new Item<string, string>("user2", "Bob Updated") 
            });

            if (btree.Find(ctx, "user2"))
            {
                var items = btree.GetValues(ctx, "user2");
                Console.WriteLine($"Updated: {items[0].Key} -> {items[0].Value}");
            }

            // 7. Remove (Delete)
            Console.WriteLine("Removing user3...");
            btree.Remove(ctx, new List<string> { "user3" });

            if (!btree.Find(ctx, "user3"))
            {
                Console.WriteLine("user3 removed successfully.");
            }

            // 8. Commit
            trans.Commit();
            Console.WriteLine("Transaction committed.");
        }
        catch (Exception ex)
        {
            trans.Rollback();
            Console.WriteLine($"Error: {ex.Message}");
        }
    }
}

File: bindings/java/src/main/java/com/sharedcode/sop/examples/BTreeBasic.java
package com.sharedcode.sop.examples;

import com.sharedcode.sop.*;
import java.util.Collections;
import java.util.List;

public class BTreeBasic {
    public static void run() {
        System.out.println("\n--- Running Basic B-Tree Operations ---");

        try (Context ctx = new Context()) {
            // 1. Open Database
            DatabaseOptions dbOpts = new DatabaseOptions();
            dbOpts.stores_folders = Collections.singletonList("sop_data_basic");
            // Assuming 0 is Standalone, similar to C# default or explicit enum
            dbOpts.type = DatabaseType.Standalone; 
            
            Database db = new Database(dbOpts);

            // 2. Start Transaction
            try (Transaction trans = db.beginTransaction(ctx)) {
                // 3. Create/Open B-Tree
                // Note: Java generics are erased, so we pass class tokens
                BTree<String, String> btree = db.newBtree(ctx, "users_basic", trans, null, String.class, String.class);

                // 4. Add Items (Create)
                System.out.println("Adding users...");
                btree.add(new Item<>("user1", "Alice"));
                btree.add(new Item<>("user2", "Bob"));
                btree.add(new Item<>("user3", "Charlie"));

                // 5. Find & Get (Read)
                if (btree.find("user1")) {
                    List<Item<String, String>> items = btree.getValues(Collections.singletonList(new Item<>("user1", null)));
                    if (items != null && !items.isEmpty()) {
                        System.out.println("Found: " + items.get(0).key + " -> " + items.get(0).value);
                    }
                }

                // 6. Update
                System.out.println("Updating user2...");
                btree.update(Collections.singletonList(new Item<>("user2", "Bob Updated")));

                if (btree.find("user2")) {
                    List<Item<String, String>> items = btree.getValues(Collections.singletonList(new Item<>("user2", null)));
                    if (items != null && !items.isEmpty()) {
                        System.out.println("Updated: " + items.get(0).key + " -> " + items.get(0).value);
                    }
                }

                // 7. Remove (Delete)
                System.out.println("Removing user3...");
                btree.remove("user3");

                if (!btree.find("user3")) {
                    System.out.println("user3 removed successfully.");
                }

                // 8. Commit
                trans.commit();
                System.out.println("Transaction committed.");
            } catch (Exception ex) {
                System.out.println("Error: " + ex.getMessage());
                ex.printStackTrace();
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
    
    public static void main(String[] args) {
        run();
    }
}
